packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,135 @@
+    ///         permanently locked. The same will occur if the target on the other chain is
+    ///         considered unsafe (see the _isUnsafeTarget() function).
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _target      Target contract or wallet address.
+    /// @param _message     Message to trigger the target address with.
+    function sendMessage(uint256 _destination, address _target, bytes calldata _message) external payable {
+        require(_destination != block.chainid, "L2ToL2CrossDomainMessenger: cannot send message to self");
+
+        bytes memory data = abi.encodeCall(
+            L2ToL2CrossDomainMessenger.relayMessage,
+            (_destination, messageNonce(), msg.sender, _target, msg.value, _message)
+        );
+        emit SentMessage(_destination, _target, _message, data);
+        msgNonce++;
+    }
+
+    /// @notice Relays a message that was sent by the other CrossDomainMessenger contract. Can only
+    ///         be executed via cross-chain call from the other messenger OR if the message was
+    ///         already received once and is currently being replayed.
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _nonce       Nonce of the message being relayed.
+    /// @param _sender      Address of the user who sent the message.
+    /// @param _target      Address that the message is targeted at.
+    /// @param _value       ETH value to send with the message.
+    /// @param _message     Message to send to the target.
+    function relayMessage(
+        uint256 _destination,
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        uint256 _value,
+        bytes memory _message
+    )
+        external
+        payable
+    {
+        require(msg.sender == crossL2Inbox, "L2ToL2CrossDomainMessenger: sender not CrossL2Inbox");
+        require(
+            CrossL2Inbox(crossL2Inbox).origin() == address(this),
+            "L2ToL2CrossDomainMessenger: CrossL2Inbox origin not this contract"
+        );
+        require(_destination == block.chainid, "L2ToL2CrossDomainMessenger: destination not this chain");
+        require(_target != address(this), "L2ToL2CrossDomainMessenger: target not this contract");
+
+        bytes32 messageHash = keccak256(abi.encode(_destination, _nonce, _sender, _target, _value, _message));
+        require(successfulMessages[messageHash] == false, "L2ToL2CrossDomainMessenger: message already relayed");
+
+        assembly {
+            tstore(CROSS_DOMAIN_MESSAGE_SENDER_SLOT, _sender)
+        }
+
+        bool success = SafeCall.callWithAllGas({ _target: _target, _value: _value, _calldata: _message });
+
+        if (success) {
+            // This check is identical to one above, but it ensures that the same message cannot be relayed
+            // twice, and adds a layer of protection against rentrancy.
+            assert(successfulMessages[messageHash] == false);
```

Comment:
Reentrency is impossible given that this can only be called by the cross l2 inbox which has 1 function that is only eoa